[Testing] Stress / load test ephemeral & persistent chains

red-0ne opened the pull request:
## Summary

Make the necessary adjustments to the load test framework to be able to run on persistent chains.

## Issue

The load testing framework was not able to run without scaling `Gateways` and `Suppliers`. Which is not the goal for testing persistent chains like DevNet and TestNet.

- #531 

## Type of change

Select one or more:

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

**Documentation changes** (only if making doc changes)
- [ ] `make docusaurus_start`; only needed if you make doc changes

**Local Testing** (only if making code changes)
- [ ] **Unit Tests**: `make go_develop_and_test`
- [ ] **LocalNet E2E Tests**: `make test_e2e`
- See [quickstart guide](https://dev.poktroll.com/developer_guide/quickstart) for instructions

**PR Testing** (only if making code changes)
- [ ] **DevNet E2E Tests**: Add the `devnet-test-e2e` label to the PR.
    - **THIS IS VERY EXPENSIVE**, so only do it after all the reviews are complete.
    - Optionally run `make trigger_ci` if you want to re-trigger tests without any code changes
    - If tests fail, try re-running failed tests only using the GitHub UI as shown [here](https://github.com/pokt-network/poktroll/assets/1892194/607984e9-0615-4569-9452-4c730190c1d2)


## Sanity Checklist

- [x] I have tested my changes using the available tooling
- [x] I have commented my code
- [x] I have performed a self-review of my own code; both comments & source code
- [x] I create and reference any new tickets, if applicable
- [x] I have left TODOs throughout the codebase, if applicable

Olshansk commented on load-testing/config/load_test_manifest_reader.go:
```diff
@@ -11,15 +11,23 @@ import (
 type ProvisionedActorConfig struct {
 	// KeyName is the **name** of the key in the keyring to be used by the given actor.
 	KeyName string `yaml:"key_name"`
+	// Address is the address of the actor, which is used to identify already staked
```
What does persistent chains mean? If applicable, #PUC

Olshansk replied:
What does persistent chain mean? If applicable, #PUC

Olshansk commented on load-testing/config/load_test_manifest_reader.go:
```diff
@@ -11,15 +11,23 @@ import (
 	KeyName string `yaml:"key_name"`
+	// Address is the address of the actor, which is used to identify already staked
+	// actors in the network in persistent chains.
+	Address string `yaml:"address"`
```
Why can't we use `keyName` to determine the the address?

Olshansk commented on load-testing/tests/relays_stress.feature:
```diff
@@ -5,13 +5,9 @@ Feature: Loading gateway server with relays
     And a rate of "1" relay requests per second is sent per application
```
1. Why was the granularity of this test resolved?
2. If (1) is related to us relying on a "persistent chain" is there a way to have two tests:
- High granularity on ephemeral chain
- Low granularity on persistent chain

red-0ne replied:
Not sure if I'm understanding this right.

The RPS is not related to the type of chain tested, this number could be changed to any integer we want.

Obviously we can increase the rate to have a higher one, but since we are parsing it as integer we currently can't go below.

We can offer a lower RPS by either:
1. Parse the input number as a float
2. Change the step to something like `a rate of "<int>" relay requests per "<int> seconds is sent per application`

WDYT?

Olshansk commented on load-testing/config/load_test_manifest_reader.go:
```diff
@@ -11,15 +11,23 @@ import (
 	KeyName string `yaml:"key_name"`
+	// Address is the address of the actor, which is used to identify already staked
+	// actors in the network in persistent chains.
+	Address string `yaml:"address"`
```
Why can't we use `keyName` to determine the the address?

1. We already have helpers for this.
2. I feel like this creates two sources of truth.
3. I feel like we're getting into a weird scenario where there's implicit institutional knowledge

red-0ne replied:
We can't use the `keyName` in that case because the `keyName->address` association comes from the fact that we have that key present the test environment's `keyRing`. Which is not the case for TestNet where `Gateways` are owned and operated by other's and we don't have their keys in our kyring.

I agress that there's no explicit indications on whether to provide `address`es or `keyNames` and feel like comments are not enough.

An alternative would be to have different manifests for each case. It still remains that the `is_ephemeral` flag is the one that dictates whether the test targets a `LocalNet` or a `TestNet`. Rather than the filename.

We may also have dedicated manifest parsers for clarity too.

Olshansk replied:
Can you #PUC with a a short TODO_IMPROVE, and link to this thread (https://github.com/pokt-network/poktroll/pull/540/files#r1607291985) as well?

Olshansk commented on load-testing/config/load_test_manifest_reader.go:
```diff
@@ -11,15 +11,23 @@ import (
-	Gateways  []ProvisionedActorConfig `yaml:"gateways"`
+	// PersistentChain is a flag that indicates whether the test is expected to be
+	// run on localnet or long living chains (i.e. TestNet/DevNet).
+	PersistentChain bool                     `yaml:"persistent_chain"`
```
I'm open to having a discussion about this because it'll involve updating the whole PR, but I'd rather we switch the whole thing from `PersistentChain` to `EphemeralChain`.

- I think the default is that chains are persistent.
- When I read this, I was confused by what a persistent chain is.
- I'd argue that "ephemeral chain" is a lot more clear than "persistent chain" and self explanatory of the alternatives.

red-0ne replied:
I prefer `ephemeral chain` too, I'll update the code.

Olshansk commented on load-testing/config/load_test_manifest_reader.go:
```diff
@@ -11,15 +11,23 @@ import (
-	Gateways  []ProvisionedActorConfig `yaml:"gateways"`
+	// PersistentChain is a flag that indicates whether the test is expected to be
+	// run on localnet or long living chains (i.e. TestNet/DevNet).
+	PersistentChain bool                     `yaml:"persistent_chain"`
```
Add `Is` prefix to w/e the final variable ends up being.

Olshansk commented on load-testing/config/load_test_manifest_reader.go:
```diff
@@ -72,5 +78,23 @@ func ParseLoadTestManifest(manifestContent []byte) (*LoadTestManifestYAML, error
 		}
 	}
 
+	if parsedManifest.PersistentChain {
```
For ephemeral chains (non persistent chains), we need to explain why we don't configure suppliers.

#PUC

Olshansk commented on load-testing/tests/init_load_test.go:
```diff
@@ -11,12 +11,19 @@ import (
+	flag.StringVar(
+		&flagManifestFilePath,
+		"manifest",
+		"../../loadtest_manifest.yaml",
```
You assigned a default but not using it.

1. Move the default to a constant
2. Don't assign the default to the var
3. Use the constant here

bryanchriswhite replied:
@Olshansk @red-0ne I would argue that it's be more idiomatic to construct the path with `filepath.Join`. This would also be OS-agnostic.

Olshansk replied:
I can't argue with what @bryanchriswhite is arguing for.

Olshansk commented on load-testing/tests/relays_stress_helpers_test.go:
```diff
@@ -48,9 +48,10 @@ import (
+	apps            actorLoadTestIncrementPlan
+	gateways        actorLoadTestIncrementPlan
+	suppliers       actorLoadTestIncrementPlan
+	persistentChain bool
```
`isEphemeralChain` would work better

Olshansk commented on loadtest_manifest.yaml:
```diff
@@ -1,6 +1,10 @@
 # List of pre-provisioned suppliers used for load testing
 # The number of pre-provisioned **LocalNet** actors are managed in
 # localnet_config.yaml by the respective actors `count` property.
+persistent_chain: false
```
`is_ephemeral_chain: true` IMO sounds better.

Olshansk commented on loadtest_manifest.yaml:
```diff
@@ -22,4 +26,15 @@ gateways:
\ No newline at end of file
+    exposed_url: http://localhost:42081
+
+# Persistent chain manifest example:
```
Why can't we enable this in this PR?

red-0ne replied:
Since the current manifest is used to run the test. The `TestNet` and `LocalNet` scenarios are mutually exclusive, so only one of them could be uncommented/activated.

As mentioned in previous comments, I suggest having a second manifest that exercises testing non-ephemeral chains. We'll indicate which one to trigger by pointing to the corresponding manifest file.

WDYT?

Olshansk commented on load-testing/tests/relays_stress_helpers_test.go:
```diff
@@ -115,22 +116,33 @@ func (s *relaysSuite) initFundingAccount(fundingAccountKeyName string) {
+		// In the case of persistent chain, the test has no entry for the gateway
+		// and no keyName is provided. The gateway address is used to identify the
+		// gateway in the network.
+		if s.persistentChain {
```
1. The code is going to get really messy with all of these ifs all over the place
2. There is already an implicit assumption that we should only define `Address` or `KeyName` but not both
3. I think this means we may need:
- Mutiple structs
- Multiple functions

It gets messy when you star doing this (and then adding another bool, and another)

red-0ne replied:
We considered having multiple functions but they will be almost duplicated (except for what's in these conditionals).

I propose to keep but document them. And refactor if/when we introduce more functionality in the future.

Olshansk commented on load-testing/tests/relays_stress_helpers_test.go:
```diff
@@ -457,6 +482,30 @@ func (s *relaysSuite) mapStakingInfoWhenStakingAndDelegatingNewApps(
+		applications = append(applications, application)
+	}
+
+	if s.persistentChain {
```
Anywhere we return early on `persistentChain` (or potentially epehmeralChain), we need to add a comment why.

Olshansk reviewed (changes requested):
DId a first round of reviews but think we need to iterate a bit.

1. I think `persistent_chain` should be renamed to `is_ephemeral_chain` which would require negating all the ifs & updating lots of comments.

2. Anytime we return early based on the value on the bool, we need to #PUC explaining why.

3. The `keyName` / `address` feels like there's an implicit `oneof` that's not enforced by comments, or the reader, or an explanation, etc... I think there's an opportunity to create structures for it, but am not sure what the simplest path forward is. The goal is to remove the implicit institutional knowledge added in the PR.

Olshansk commented:
@red-0ne Please re-request review and update PR title when this is ready.

red-0ne commented:
* Separated (non-)ephemeral config parsing validation.
* Now we have different manifest for ephemeral and non-ephemeral tests.
* Added Makefile targets for each of the ephemeral and non-ephemeral tests.
* Moved the manifest from the root directory to `load-testing`

Possible improvement:

Be able to specify the test plan (feature files) to execute from the manifest file

Olshansk commented on load-testing/config/load_test_manifest_reader.go:
```diff
@@ -11,15 +11,23 @@ import (
 type LoadTestManifestYAML struct {
-	Suppliers []ProvisionedActorConfig `yaml:"suppliers"`
-	Gateways  []ProvisionedActorConfig `yaml:"gateways"`
+	// IsEphemeralChain is a flag that indicates whether the test is expected to be
+	// run on localnet or long living chains (i.e. TestNet/DevNet).
```
```suggestion
	// IsEphemeralChain is a flag that indicates whether the test is expected to
	// run on LocaLnet or a long-living remote chain (i.e. TestNet/DevNet).
```

Olshansk commented on load-testing/config/load_test_manifest_reader.go:
```diff
@@ -36,41 +44,95 @@ func ParseLoadTestManifest(manifestContent []byte) (*LoadTestManifestYAML, error
-		if supplier.KeyName == "" {
-			return nil, ErrLoadTestInvalidManifest.Wrap("empty supplier key name")
+	for _, supplier := range manifest.Suppliers {
+		if len(supplier.KeyName) == 0 {
```
Why don't we check for `supplier.Address`?

Olshansk commented on load-testing/config/load_test_manifest_reader.go:
```diff
@@ -36,41 +44,95 @@ func ParseLoadTestManifest(manifestContent []byte) (*LoadTestManifestYAML, error
+			return nil, ErrEphemeralChainLoadTestInvalidManifest.Wrap("empty gateway key name")
+		}
+
+		if len(gateway.Address) > 0 {
```
#PUC why this is forbidden

Olshansk commented on load-testing/config/load_test_manifest_reader.go:
```diff
@@ -36,41 +44,95 @@ func ParseLoadTestManifest(manifestContent []byte) (*LoadTestManifestYAML, error
+		return nil, ErrNonEphemeralChainLoadTestInvalidManifest.Wrap("empty gateways entry")
+	}
+
+	if len(manifest.Suppliers) > 0 {
```
#PUC for forbidden configs

Olshansk commented on load-testing/config/load_test_manifest_reader.go:
```diff
@@ -36,41 +44,95 @@ func ParseLoadTestManifest(manifestContent []byte) (*LoadTestManifestYAML, error
+	}
+
+	for _, gateway := range manifest.Gateways {
+		if len(gateway.KeyName) > 0 {
```
#PUC why

Olshansk commented on load-testing/tests/relays_stress_helpers_test.go:
```diff
@@ -115,22 +117,36 @@ func (s *relaysSuite) initFundingAccount(fundingAccountKeyName string) {
+	for _, gateway := range loadTestManifest.Gateways {
+		gatewayId := gateway.Address
+		// In the case of non-ephemeral chain load testing, the test uses the
+		// gateway's keyName as the gatewayId.
```
#PUC why

Olshansk reviewed (approved):
PTAL at the NITs I left related to the comments, but not blocking and LGTM.

Love how this turned out after the first iteration!

Olshansk commented on load-testing/tests/relays_stress_helpers_test.go:
```diff
@@ -517,29 +513,34 @@ func (s *relaysSuite) sendFundAvailableActorsTx(
-		supplier := s.addActor(keyName, supplierStakeAmount)
+	suppliersAdded := int64(0)
+	for _, supplierAddress := range s.availableSupplierAddresses {
+		if suppliersAdded > plans.suppliers.maxActorCount {
```
Not a blocker, but if we start counting at 0, we reach the max when it's `>=`, not `>`.

Ditto elsewhere.

Olshansk commented on load-testing/tests/relays_stress_helpers_test.go:
```diff
@@ -760,17 +761,16 @@ func (plan *actorLoadTestIncrementPlan) shouldIncrementSupplierCount(
+	require.NotNil(s, keyRecord)
 
 	return &accountInfo{
 		accAddress:    accAddress,
```
Do we still need both `accAddress` and `address` or is there a way to just keep one?

Olshansk commented on load-testing/tests/relays_stress_test.go:
```diff
@@ -77,8 +77,8 @@ var (
-	// fundingAccountKeyName is the key name of the account used to fund other accounts.
-	fundingAccountKeyName = "pnf"
+	// fundingAccountAddress is the key name of the account used to fund other accounts.
+	fundingAccountAddress = "pokt1eeeksh2tvkh7wzmfrljnhw4wrhs55lcuvmekkw"
```
```suggestion
	fundingAccountAddress = "pokt1eeeksh2tvkh7wzmfrljnhw4wrhs55lcuvmekkw" // address for pnf account
```

Olshansk commented on load-testing/tests/relays_stress_test.go:
```diff
@@ -217,8 +225,8 @@ type relaysSuite struct {
-	keyName       string
+	// The address of the account available in the keyring used by the test.
+	address       string
 	accAddress    sdk.AccAddress
```
We have `address` and `accAddress`. Do we need both?

red-0ne replied:
The reason why `accAddress` is kept is that `bank.NewMsgSend` and  `KeyRing.GetKeyByAddress` require `AccAddress` structs.

We can create them on the fly though. I'll give their removal a try since we can derive one from the other.

Olshansk commented:
@red-0ne Love the simplification, so much clearer now!

Only big remaining question is whether we need both `address` and `accAddress` in `accountInfo`.

I know I can pull the PR and figure it out myself but just asking to expedite things.

red-0ne commented:
@Olshansk , `accAddress` removed successfully

Olshansk reviewed (approved):
:shipit:
